### PR TITLE
Render Magic

### PR DESCRIPTION
**Pull request by RlonRyan**
Tried to simplify some rendering things, and add some useful utility methods.

Seems to work pretty well, and opens up a door to future improvements.

### CODE REVIEW

File: src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderValve.java [hunk 1]
```diff
@@ -96,185 +103,78 @@ protected boolean doWorldRender(Tessellator tessellator2, IBlockAccess world, do
                     GL11.glPopMatrix();
                 }
             } else {
-                float unit = Constants.UNIT;
                 this.renderWoodChannel(valve, tessellator2);
-                if ((!this.shouldBehaveAsTESR()) && (valve.getDiscreteScaledFluidLevel() > 0)) {
+                if ((!this.shouldBehaveAsTESR()) && (valve.getDiscreteFluidLevel() > 0)) {
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderValve.java]: This should not have changed, if there is any amount of water in the channel inf ]0, LVL1[, nothing will render, even though there will be quite a bit of water in it.
The old method rendered water at the bottom of the channel.

File: src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityChannelFull.java [hunk 2]
```diff
@@ -9,11 +10,11 @@ public TileEntityChannelFull() {
     }
 
     @Override
-    public boolean hasNeighbour(char axis, int direction) {
+    public boolean hasNeighbourCheck(ForgeDirection direction) {
         if (this.worldObj == null) {
             return true;
         } else {
-            return super.hasNeighbour(axis, direction);
+            return super.hasNeighbour(direction);
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityChannelFull.java]: should be hasNeighbourCheck, client doesn't cache the neighbours.

File: src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityChannelFull.java [hunk 3]
```diff
@@ -9,11 +10,11 @@ public TileEntityChannelFull() {
     }
 
     @Override
-    public boolean hasNeighbour(char axis, int direction) {
+    public boolean hasNeighbourCheck(ForgeDirection direction) {
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityChannelFull.java]: This needs @SideOnly(Side.CLIENT) annotation

File: src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java [hunk 4]
```diff
@@ -37,6 +37,7 @@
     private int zSize=1;
 
     //boolean to convert pre-1.4 tanks
+    //Lets remove this... it has been quite some time...
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java]: No it hasn't, you'd be amazed how many people are still on 1.3, all FTB packs still are.

File: src/main/java/com/InfinityRaider/AgriCraft/utility/RenderHelper.java [hunk 5]
```diff
@@ -48,38 +50,169 @@ public static ResourceLocation getItemResource(IIcon icon) {
...
+    		case WEST:
+    			return 'x';
+    		case UP:
+    		case DOWN:
+    			return 'y';
+    		case NORTH:
+    		case SOUTH:
+    			return 'z';
+    		default:
+    			return '?'; //Umm... Ignore this.
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/utility/RenderHelper.java]: make it return UNKNOWN

File: src/main/java/com/InfinityRaider/AgriCraft/utility/RenderHelper.java [hunk 6]
```diff
@@ -48,38 +50,169 @@ public static ResourceLocation getItemResource(IIcon icon) {
...
+     */
+    public static int getMagnitude(ForgeDirection direction) {
+    	return direction.offsetX + direction.offsetY + direction.offsetZ; // Should always be in the -1 to 1 range.
+    }
+    
+    /**
+     * Rotates a plane. This is impressively useful, but may not be impressively efficient.
+     * Always returns a 6-element array. Defaults to the north direction (the base direction).
+     * <p>
+     * This is for use up to the point that a way to rotate lower down is found. (IE. OpenGL).
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/utility/RenderHelper.java]: OpenGL calls in ISBRH requires a hacky workaround.

File: src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderBlockBase.java [hunk 7]
```diff
@@ -263,6 +267,19 @@ protected void drawScaledFaceDoubleYZ(Tessellator tessellator, float minY, float
...
     }
 
+    /**
+     * Draws a prism.
+     * <p>
+     * The prism coordinates range from 0 to {@link Constants#WHOLE}.
+     * </p><p>
+     * The prism is relative to the bottom left corner of the block.
+     * </p>
+     */
```
Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderBlockBase.java]: I like this method, even though it'll only do 90° rotations, I like it.

File: src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java (hunk 4 above)
Comment (RlonRyan) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java]: That's terrible...

File: src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderValve.java (hunk 1 above)
Comment (RlonRyan) [inline src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderValve.java]: This does render the water in that range. It just removes the unneeded multiplication.

File: src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityChannelFull.java (hunk 2 above)
Comment (RlonRyan) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityChannelFull.java]: Lol... durp.

File: src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderBlockBase.java (hunk 7 above)
Comment (RlonRyan) [inline src/main/java/com/InfinityRaider/AgriCraft/renderers/blocks/RenderBlockBase.java]: I like it too... lol.

File: src/main/java/com/InfinityRaider/AgriCraft/utility/RenderHelper.java (hunk 6 above)
Comment (RlonRyan) [inline src/main/java/com/InfinityRaider/AgriCraft/utility/RenderHelper.java]: Yeah... the rendering in minecraft has always been a mess.

**Comment by RlonRyan** [comment]
Sorry about the last PR... I wasn't entirely coherent at the time I made it.

**Comment by RlonRyan** [comment]
Also, about the color multiplier... I fail to see what impact it makes...

It still seems like we are having some major issues getting light levels to render properly...
For the most part, in my testing, the entire tank-conduit system likes to render at the same light level...

**Comment by RlonRyan** [comment]
Finally, I think we should probably making the custom wood blocks separate instances of the custom wood block class... So that we aren't constantly setting and resetting the parameters on the the block, which could remain constant. This would also make material differences easier, allowing for iron channels?

**Comment by InfinityRaider** [comment]
But each block has its own seperate instance of BlockCustomwood.
Also, without CM:
http://puu.sh/iBx7k/908e7e6e70.jpg
with CM: http://puu.sh/k4hBi/1e8e09ab79.jpg
See how it smoothly transistions from the oak planks to the tank?

**Comment by RlonRyan** [comment]
Ah. Nicely done!

**Comment by RlonRyan** [comment]
Ok, if the instances are separate, it looks like I just need to change some constructors then.

**Comment by InfinityRaider** [comment]
I'll merge this tomorrow after my lecture.

**Comment by RlonRyan** [comment]
Sounds good!

Have fun at the lecture! (Whatever it is..?)

**Comment by RlonRyan** [comment]
FYI, I might change the way valves render in the inventory by then, as ATM they look kinda confusing.

File: src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java (hunk 4 above)
Comment (RlonRyan) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java]: I guess the goal is to do well enough in this latest release to get the ftb people to update, and then we can consider the majority updated, so we can finally clean out old stuff?

Comment (InfinityRaider) [inline src/main/java/com/InfinityRaider/AgriCraft/tileentity/irrigation/TileEntityTank.java]: I want to finish the rendering on the peripheral and hten everything I want for the next release is in.

**Comment by RlonRyan** [comment]
Does this look O.K.?

**Comment by RlonRyan** [comment]
Tell me if I should remove that last commit... I think it makes the tanks more manageable though...

It didn't seem to break anything...

**Comment by InfinityRaider** [comment]
Remove it and make a new PR for it please.

**Comment by RlonRyan** [comment]
Got it!

**Pull request merged**